File: R/denoising.R
##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Image denoising by NLM Algorithm
##' @param img rip object of 1, 2 or 3 channels
##' @param h  Parameter regulating filter strength. 
##' @param templeteWindowSize Size in pixels of the template patch that is used to compute weights. Should be odd
##' @param searchWindowSize Size in pixels of the window that is used to compute weighted average for given pixel. Should be odd.
##' @return Denoised rip object with the same size as \code{img}
##' @author Kaustav Nandy
fastNlMeansDenoising <- function(img, h, templeteWindowSize, searchWindowSize)
{
    mdenoise <- Module("denoise", PACKAGE = "RIPOpenCV")
    mdenoise$nlmeansDenoising(img,  h, templeteWindowSize, searchWindowSize)
}
##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Image denoising by NLM Algorithm for colored image
##' @param img rip object of 3 channels
##' @param h Parameter regulating filter strength. 
##' @param templeteWindowSize Size in pixels of the template patch that is used to compute weights. Should be odd
##' @param searchWindowSizeSize in pixels of the window that is used to compute weighted average for given pixel. Should be odd. 
##' @return Denoised rip object with the same size as \code{img}
##' @author Kaustav Nandy
fastNlMeansDenoisingColored <- function(img, h, templeteWindowSize, searchWindowSize)
{
    mdenoise <- Module("denoise", PACKAGE = "RIPOpenCV")
    mdenoise$nlmeansDenoisingCol(img,  h, templeteWindowSize, searchWindowSize)
}

File: src/denoising.cpp
#include "conversion.h"

using namespace cv;
using namespace Rcpp;

Rcpp::NumericMatrix nlmeansDenoising(Rcpp::NumericMatrix imgMat, float h, 
				     int templeteWindowSize, int searchWindowSize)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 0);
    cv::Mat outImg;
    cv::fastNlMeansDenoising(M, outImg, h, templeteWindowSize, searchWindowSize);
    return (convertMat_CV2RCPP(outImg));
}


Rcpp::NumericMatrix nlmeansDenoisingCol(Rcpp::NumericMatrix imgMat, float h, 
					int templeteWindowSize, int searchWindowSize)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 0);
    cv::Mat outImg;
    cv::fastNlMeansDenoisingColored(M, outImg, h, templeteWindowSize, searchWindowSize);
    return (convertMat_CV2RCPP(outImg));
}

RCPP_MODULE(denoise) 
{
    function("nlmeansDenoising", &nlmeansDenoising, "NL means denoising");
    function("nlmeansDenoisingCol", &nlmeansDenoisingCol, "NL means denoising");
}
